A0-2760: Support for displaying future committies

Marcin-Radecki opened the pull request:
Changes in two Places of Page Staking that supports future committees:
1. In Performance tab, user can now go to further session than a current session, but no more than last session of the current era. When choosing such future session, a future block committee is shown to user. Data is taken from `alepSessionApi.predictSessionCommittee` call. 
Minor change from UX perspective is that once Performance page loads current session, it stays that way, even when new session starts. This assumption makes code more readable and less coupled - I refactored this page pretty much which should be better now to maintain, as it was a bit messy. If we want to keep current behaviour please let me know - I'd need to add some additional UX element that preserves this "Current session mode".

2. In Validator stats tab, a new toggle is added which gives user two choices query past performance (current behavior) or all future session in the current era which validator will be in a block production commitee. This reuses the same hook for querying data as point 1.

3. New tab  `Future Committee` in Staking app. It displays all era validators in a table, with information about next future committee.

Marcin-Radecki committed:
First working version - changes to support future committies.

Marcin-Radecki committed:
Yarn lint

Marcin-Radecki committed:
Changes to Staking/Query tab

Marcin-Radecki committed:
yarn lint

youPickItUp commented on packages/apps-config/src/types.ts:
```diff
@@ -13,6 +14,10 @@ declare module '@polkadot/api-base/types/calls' {
        * The API to query account nonce (aka transaction index)
        **/
       yearlyInflation?: AugmentedCall<ApiType, () => Observable<Perbill>>;
+      /**
+      Predict finality and block production committee
+       **/
+      predictSessionCommittee?: AugmentedCall<ApiType, (session: u32) => Observable<string>>;
```
I think the exact type should be
```jsx
import type { ApiTypes, AugmentedCall, DecoratedCallBase } from '@polkadot/api-base/types';
import type { Struct, Vec } from '@polkadot/types';
import type { SessionIndex } from '@polkadot/types/interfaces';
import type { AccountId32, Perbill } from '@polkadot/types/interfaces/runtime';
import type { AnyNumber, Codec, Observable } from '@polkadot/types/types';
import type { Enum, Result, u8 } from '@polkadot/types-codec';

interface SessionCommittee<T extends Codec> extends Struct {
  readonly finalizers: Vec<T>;
  readonly producers: Vec<T>;
}

interface SessionNotWithinRange extends Struct {
  readonly lowerLimit: SessionIndex;
  readonly upperLimit: SessionIndex;
}
interface SessionValidatorError extends Enum {
  readonly isOther: boolean;
  readonly asOther: Vec<u8>;
  readonly isSessionNotWithinRange: boolean;
  readonly asSessionNotWithinRange: SessionNotWithinRange;
  readonly type: 'Other' | 'SessionNotWithinRange';
}
declare module '@polkadot/api-base/types/calls' {
  interface AugmentedCalls<ApiType extends ApiTypes> {
    /** 0xbc9d89904f5b923f/1 */
    alephSessionApi?: {
      /**
       * The API to query account nonce (aka transaction index)
       **/
      yearlyInflation?: AugmentedCall<ApiType, () => Observable<Perbill>>;
      /**
      Predict finality and block production committee
       **/
      predictSessionCommittee?: AugmentedCall<ApiType, (session: SessionIndex | AnyNumber | Uint8Array) => Observable<Result<SessionCommittee<AccountId32>, SessionValidatorError>>>;
      /**
       * Generic call
       **/
      [key: string]: DecoratedCallBase<ApiType> | undefined;
    };
  }
}
```
It would make parsing done in `useFutureSessionCommittee` simpler.

Disclaimer: it's not generated, I've tried to map the backend type 1 to 1, it seems fine, but please test it before you use it as I could have missed / misunderstood something.

Marcin-Radecki replied:
Good point, much appreciated for perfect snippet. This should be now addressed.

youPickItUp committed:
A0-2760: Style "Validator stats" toggles (#138)

Marcin-Radecki committed:
Review remarks. Fixed a few bugs and found one existing bug and not fixed it.

Marcin-Radecki committed:
Added support for current mode to display next committee session

Marcin-Radecki commented:
@youPickItUp  could you help me with style, I cannot make those two columns aligned, this is when the new toggle `display new committee` is set in Staking / Performance tab :pray: 

![image](https://github.com/user-attachments/assets/cfa3e609-5616-4a1e-b855-2b16490d3777)

Marcin-Radecki commented:
> @youPickItUp could you help me with style, I cannot make those two columns aligned, this is when the new toggle `display new committee` is set in Staking / Performance tab 🙏
> 
> ![image](https://private-user-images.githubusercontent.com/3909333/409997824-cfa3e609-5616-4a1e-b855-2b16490d3777.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mzg3NjExNTYsIm5iZiI6MTczODc2MDg1NiwicGF0aCI6Ii8zOTA5MzMzLzQwOTk5NzgyNC1jZmEzZTYwOS01NjE2LTRhMWUtYjg1NS0yYjE2NDkwZDM3NzcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MDIwNSUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTAyMDVUMTMwNzM2WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NDQzY2JhNDQwZTlhNjZlMjY1ODdiYjM4MmNkY2YzYzkzN2VlNmM1ZTEwYjJiZTVkOGEwNWY2NmQ4ZTAxYjEyNiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.aBNUKj6JtL2GDImXDXsCN14TRHpgmbtSkqfiJBcy4rA)

Actually, please ignore. I'll need to rework this anyway

Marcin-Radecki committed:
Revert "Added support for current mode to display next committee session"

This reverts commit c1c208d5b34f5575f6de1f0adc766a6e1e274183.

Fixed one issue in suspensions tab. Fixed another issue in Staking / Performance tab in which not all validators were
returned.

Marcin-Radecki committed:
Added Future Committee tab

Marcin-Radecki committed:
Added info when there are no committiees in the current era

Marcin-Radecki commented:
@youPickItUp this PR is ready for review

youPickItUp commented on packages/page-staking/src/FutureCommittee/index.tsx:
```diff
@@ -0,0 +1,47 @@
+import FutureValidators from './FutureValidators.js';
+import SummarySession from './SummarySession.js';
+
+function PerformancePage (): React.ReactElement {
```
Wrong name

Marcin-Radecki replied:
Good point, fixed

youPickItUp commented on packages/page-staking/src/FutureCommittee/FutureValidators.tsx:
```diff
@@ -0,0 +1,124 @@
+          nextSessionInCommittee: nextCommittee?.session
+        };
+      }).sort((_, b) => {
+        return b.currentSessionCommittee ? 1 : -1;
```
How about:
```ts
return Number(b.currentSessionCommittee) - Number(a.currentSessionCommittee);
```

Theoretically here are [requirements a comparator function must fulfill](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort#description). I think current solution might work, but I'm not sure :sweat_smile: 

>    Pure: The comparator does not mutate the objects being compared or any external state. (This is important because there's no guarantee when and how the comparator will be called, so any particular call should not produce visible effects to the outside.)

>    Stable: The comparator returns the same result with the same pair of input.

>    Reflexive: compareFn(a, a) === 0.

>    Anti-symmetric: compareFn(a, b) and compareFn(b, a) must both be 0 or have opposite signs.

>    Transitive: If compareFn(a, b) and compareFn(b, c) are both positive, zero, or negative, then compareFn(a, c) has the same positivity as the previous two.

Marcin-Radecki replied:
On the other hand - while current version might not be ideal from aesthetic point of view, proposed version relies a bit of JavaScript shenanigans , at least in my opinion. I'd leave as it is then, as I prefer readability factor here.

youPickItUp replied:
the current function doesn't fulfill the requirements for `sort` comparator function. I'm a bit worried that there might be some edge case here breaking something, though the chances are slim since it has been tested. I'm fine with current solution, if you prefer it, but the only shenanigan I see in my solution is conversion from boolean to number, which I think isn't that uncommon? Perhaps we could make is even clearer by doing:
```jsx
const currentSessionCommitteeToNumber= b.currentSessionCommittee ? 1 : 0;
```
Anyway, up to you!

Marcin-Radecki replied:
You're right current version is not handling second argument correctly. Changed to `return Number(b.currentSessionCommittee) - Number(a.currentSessionCommittee);`

youPickItUp commented on packages/page-staking/src/Performance/BlockProductionCommitteeList.tsx:
```diff
@@ -26,7 +26,7 @@ function getFiltered (displayOnlyCommittee: boolean, eraValidatorPerformances: E
 export function calculatePercentReward (blocksCreated: number | undefined, blocksTargetValue: number | undefined, lenientThresholdPercentage: number | undefined, isCommittee: boolean) {
   if (blocksCreated === undefined || blocksTargetValue === undefined || lenientThresholdPercentage === undefined) {
-    return '';
+    return '0.0';
```
I think empty is better placeholder while we're loading the data, but :shrug:

Marcin-Radecki replied:
I changed it to be consistent in all tables in Staking tab: In `Validator Stats`, `Performace` and `Future Committee`. Without this change, some of the tables have empty, and some have `0.0` value.

youPickItUp commented on packages/page-staking/src/Performance/Performance.tsx:
```diff
@@ -17,17 +16,12 @@ import BlockProductionCommitteeList from './BlockProductionCommitteeList.js';
 }
 
-function Performance ({ era }: Props): React.ReactElement<Props> {
+function Performance (): React.ReactElement<null> {
```
Return type can be omitted and will be inferred from function body
```suggestion
function Performance (): {
```

Marcin-Radecki replied:
good point, changed

youPickItUp commented on packages/page-staking/src/Performance/Summary.tsx:
```diff
@@ -35,11 +35,13 @@ function Summary (
+            <CardSummary label={t('expected block count')}>
+              <span className={expectedBlockCount ? '' : '--tmp'}>
+                {formatNumber(expectedBlockCount || 0)}
+              </span>
```
since we check for `undefined` before we could do
```suggestion
              <span>
                {formatNumber(expectedBlockCount)}
              </span>
```

Marcin-Radecki replied:
done

youPickItUp commented on packages/page-staking/src/Performance/SummarySession.tsx:
```diff
@@ -1,25 +1,46 @@
+    }
+
+    return undefined;
+  }, [eraSessionBoundaries]);
```
this could be
```suggestion
  const era =  eraSessionBoundaries?.era;
```

Marcin-Radecki replied:
good point

youPickItUp commented on packages/page-staking/src/Performance/index.tsx:
```diff
@@ -69,48 +75,67 @@ function PerformancePage (): React.ReactElement {
-          value={groupIndex}
-        />
+        <>
+          {performanceTabMode !== PerformanceTabMode.Future &&
```
we could disable unavailable buttons and show the menu in all tabs e.g.
```ts
  const groups = [
    { text: t('Era validators'), value: 'validators', isDisabled: performanceTabMode === PerformanceTabMode.Future },
    { text: t('Block production committee'), value: 'block' },
    { text: t('Finality committee'), value: 'finality', isDisabled: performanceTabMode === PerformanceTabMode.Future }
  ];
```

Marcin-Radecki replied:
This is definitely a better UX, thanks!

youPickItUp commented on packages/page-staking/src/Performance/useEraSessionBoundaries.tsx:
```diff
@@ -0,0 +1,60 @@
+  eraEndSession?: number;
+}
+
+// Returns given era start first end last session. If session is undefined, assume it's current session.
```
```suggestion
// Returns given era start first end last session.
```

Marcin-Radecki replied:
good catch

youPickItUp commented on packages/page-staking/src/Performance/useEraValidators.tsx:
```diff
@@ -16,9 +16,18 @@ export const useEraValidators = (session: number): string[] | undefined => {
+            .catch(console.error);
+        }
+      }).catch(console.error);
+    }).catch(console.error);
```
I think you could run `getBlock` without `hash` argument to get best block as well. I wonder if getting current session and comparing it to `session` parameter wouldn't be simpler

Marcin-Radecki replied:
1. With regards to `getBlock` good point, I would do that
2.  Unfortunately comapring to `session`  won't work, this was my first take here - for a moment I received `failed to obtain block` error in the console which was not a genuine error, but a 1 to 2 seconds time window in which genuinely there was not yet the first session block imported. This situation happened when new session happens and user is still on the page which uses `useEraValidators` hook. Not that common scenario, but still it is an issue we need to address. Agree the fix for it seems complicated, but I could not find simpler solution that worked.

Marcin-Radecki committed:
Review rework

Marcin-Radecki committed:
Deduplicate session.validators (#139)

This change just deduplicates `session.validators` storage. It is
required in v15 node, as that storage has 900 entries now - representing
900 slots. This PR does not affect v14 version, as deduplication does
not affect `session.validators` order.

Implementation is not ideal, as still there's some code duplication -
ideally I'd like to see `api.query.session.vliadators` only once in
`page-staking`.

Also, I analyzed if those v15 changes can impact other part of this
repo, but they are not:
* apart `page-staking` there are no other usages of
`api.query.session.vliadators`
* `api-derive` uses `api.query.session.vliadators`, but thankfully we
don't use any of such calls

youPickItUp commented on packages/page-staking/src/Performance/useFutureSessionCommittee.tsx:
```diff
@@ -0,0 +1,125 @@
+
+function instanceOfSessionNotWithinRange (object: any): object is SessionNotWithinRange {
+  return 'lowerLimit' in object && 'upperLimit' in object;
+}
```
more type safe option:
```jsx
function instanceOfSessionCommitteeV15 (object: SessionCommitteeV15<AccountId32> | SessionCommitteeV14<AccountId32>): object is SessionCommitteeV15<AccountId32> {
  return 'finalizers' in object && 'producers' in object;
}

function instanceOfSessionCommitteeV14 (object: SessionCommitteeV15<AccountId32> | SessionCommitteeV14<AccountId32>): object is SessionCommitteeV14<AccountId32> {
  return 'finalityCommittee' in object && 'blockProducers' in object;
}

function instanceOfSessionNotWithinRange (object: FutureCommitteeResult | SessionNotWithinRange): object is SessionNotWithinRange {
  return 'lowerLimit' in object && 'upperLimit' in object;
}
```

Marcin-Radecki replied:
good point

youPickItUp commented on packages/page-staking/src/Performance/useFutureSessionCommittee.tsx:
```diff
@@ -0,0 +1,125 @@
+function useFutureSessionCommitteeImpl (sessions: number[]): FutureCommittee[] {
+  const { api } = useApi();
+
+  const [allPredictedCommettees, setAllPredictedCommettees] = useState<(FutureCommitteeResult | undefined | SessionNotWithinRange)[]>([]);
```
Spelling: ...Comm **i** ttees

Marcin-Radecki replied:
https://en.wikipedia.org/wiki/Halley%27s_Comet ;)

changed

youPickItUp commented on packages/page-staking/src/Performance/useFutureSessionCommittee.tsx:
```diff
@@ -0,0 +1,125 @@
+    const predictSessionCommittee = api.call?.alephSessionApi?.predictSessionCommittee;
+
+    if (!predictSessionCommittee) {
+      console.error('api.call.alephSessionApi.predictSessionCommittee is undefined!');
```
perhaps we could do a `return` here?

Marcin-Radecki replied:
yes, good point

youPickItUp commented on packages/page-staking/src/Performance/useFutureSessionCommittee.tsx:
```diff
@@ -0,0 +1,125 @@
+      console.error('api.call.alephSessionApi.predictSessionCommittee is undefined!');
+    }
+
+    const predictCommitteePromises = sessions.map((session) => predictSessionCommittee?.(session as unknown as u32));
```
`as unknown as u32` isn't needed anymore

Marcin-Radecki replied:
true

youPickItUp reviewed (approved):
LGTM!

Marcin-Radecki committed:
Review remarks round 2

Marcin-Radecki committed:
Typo
